File: lib/commons/src/main/java/io/github/jinganix/webpb/commons/UrlSegment.java
/*
 * Copyright (c) 2020 The Webpb Authors, All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * https://github.com/jinganix/webpb
 */

package io.github.jinganix.webpb.commons;

/** Path param captured from url. */
public class UrlSegment {

  private final boolean accessor;

  private final String prefix;

  private final String key;

  private final String value;

  /**
   * Construct a {@link UrlSegment}.
   *
   * @param prefix prefix of the segment
   * @param key key of the param
   * @param value value of the segment
   */
  public UrlSegment(String prefix, String key, String value) {
    this.prefix = prefix;
    this.key = key;
    this.accessor = value.startsWith("{") && value.endsWith("}");
    if (this.accessor) {
      this.value = value.substring(1, value.length() - 1);
    } else {
      this.value = value;
    }
  }

  /**
   * Is query.
   *
   * @return is in query
   */
  public boolean isQuery() {
    return key != null && !key.isEmpty();
  }

  /**
   * Is accessor.
   *
   * @return accessor
   */
  public boolean isAccessor() {
    return accessor;
  }

  /**
   * Get prefix.
   *
   * @return prefix
   */
  public String getPrefix() {
    return prefix;
  }

  /**
   * Get key.
   *
   * @return key
   */
  public String getKey() {
    return key;
  }

  /**
   * Get value.
   *
   * @return value
   */
  public String getValue() {
    return value;
  }
}

File: plugin/internal/commons/url_segment.go
package commons

import "strings"

// UrlSegment is a path param captured from a URL.
type UrlSegment struct {
	accessor bool
	prefix   string
	key      string
	value    string
}

// NewUrlSegment constructs a UrlSegment.
func NewUrlSegment(prefix, key, value string) UrlSegment {
	accessor := strings.HasPrefix(value, "{") && strings.HasSuffix(value, "}")
	v := value
	if accessor {
		v = value[1 : len(value)-1]
	}
	return UrlSegment{
		accessor: accessor,
		prefix:   prefix,
		key:      key,
		value:    v,
	}
}

// IsQuery reports whether the segment is in the query string.
func (s UrlSegment) IsQuery() bool {
	return s.key != ""
}

// IsAccessor reports whether the segment is a path accessor.
func (s UrlSegment) IsAccessor() bool {
	return s.accessor
}

// Prefix returns the segment prefix.
func (s UrlSegment) Prefix() string {
	return s.prefix
}

// Key returns the segment key.
func (s UrlSegment) Key() string {
	return s.key
}

// Value returns the segment value.
func (s UrlSegment) Value() string {
	return s.value
}
